testnet/stacks-node/src/nakamoto_node/sign_coordinator.rs

```diff
@@ -0,0 +1,517 @@
+///  is used by Nakamoto miners to act as the coordinator for the blocks they
+///  produce.
+pub struct SignCoordinator {
+    coordinator: FireCoordinator<Aggregator>,
+    receiver: Option<Receiver<StackerDBChunksEvent>>,
+    message_key: Scalar,
+    wsts_public_keys: PublicKeys,
+    is_mainnet: bool,
+    miners_db_config: StackerDBConfig,
+    signing_round_timeout: Duration,
+}
+
+pub struct NakamotoSigningParams {
+    /// total number of signers
+    pub num_signers: u32,
+    /// total number of keys
+    pub num_keys: u32,
+    /// threshold of keys needed to form a valid signature
+    pub threshold: u32,
+    /// map of signer_id to controlled key_ids
+    pub signer_key_ids: HashMap<u32, HashSet<u32>>,
+    /// ECDSA public keys as Point objects indexed by signer_id
+    pub signer_public_keys: HashMap<u32, Point>,
+    pub wsts_public_keys: PublicKeys,
+}
+
+impl Drop for SignCoordinator {
+    fn drop(&mut self) {
+        STACKER_DB_CHANNEL.replace_receiver(self.receiver.take().expect(
+            "FATAL: lost possession of the StackerDB channel before dropping SignCoordinator",
+        ));
+    }
+}
+
+impl From<&[NakamotoSignerEntry]> for NakamotoSigningParams {
+    fn from(reward_set: &[NakamotoSignerEntry]) -> Self {
+        let mut weight_end = 1;
+        let mut signer_key_ids = HashMap::with_capacity(reward_set.len());
+        let mut signer_public_keys = HashMap::with_capacity(reward_set.len());
+        let mut wsts_signers = HashMap::new();
+        let mut wsts_key_ids = HashMap::new();
+        for (i, entry) in reward_set.iter().enumerate() {
+            let signer_id = u32::try_from(i).expect("FATAL: number of signers exceeds u32::MAX");
+            let ecdsa_pk = ecdsa::PublicKey::try_from(entry.signing_key.as_slice())
+                .map_err(|e| format!("Failed to convert signing key to ecdsa::PublicKey: {e}"))
+                .unwrap_or_else(|err| {
+                    panic!("FATAL: failed to convert signing key to Point: {err}")
+                });
+            let signer_public_key = Point::try_from(&Compressed::from(ecdsa_pk.to_bytes()))
+                .map_err(|e| format!("Failed to convert signing key to wsts::Point: {e}"))
+                .unwrap_or_else(|err| {
+                    panic!("FATAL: failed to convert signing key to Point: {err}")
+                });
+
+            signer_public_keys.insert(signer_id, signer_public_key);
+            let weight_start = weight_end;
+            weight_end = weight_start + entry.weight;
+            let key_ids: HashSet<u32> = (weight_start..weight_end).collect();
+            for key_id in key_ids.iter() {
+                wsts_key_ids.insert(*key_id, ecdsa_pk.clone());
+            }
+            signer_key_ids.insert(signer_id, key_ids);
+            wsts_signers.insert(signer_id, ecdsa_pk);
+        }
+
+        let num_keys = weight_end - 1;
+        let threshold = (num_keys * 70) / 100;
```

Comment:
Isn't this supposed to be `(num_keys * 700) / 1000` now?

Reply:
Yes, that sounds right. I think that this code should be deduped with the signer binary somehow. I'll try to see what I can do in that respect.

Reply:
The signer actually converts to floats and then rounded -- I refactored this so that the signer and this method both share this calculation via libsigner.